packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx

```diff
@@ -1,102 +1,141 @@
+        (packSize.isPlaceholder && packSizes.length === 1) ||
+        // Is not on hold, expired has available stock and is not a placeholder..
+        (!packSize.isPlaceholder &&
+          !packSize.isOnHold &&
+          !packSize.isExpired &&
+          packSize.hasAvailableStock)
+    ),
+    'packSize'
   );
 
-  useEffect(() => {
-    // When selected is null, set a default value - either
-    // 'any' when there are multiple unique pack sizes
-    // in the set of options, or the only option if there is only
-    // one.
+  // Add the any option when:
+  // - There are multiple valid pack sizes to choose from.
+  // - There is an expired line.
+  // - There are no valid options (i.e. there are no stock lines, only a placeholder).
+  const hasExpiredLine = packSizes.some(({ isExpired }) => isExpired);
+  if (validPackSizes.length !== 1 || hasExpiredLine) {
+    validPackSizes.unshift(createAnyOption(t)());
+  }
 
-    if (selected) return;
-    let selectedPackSize = ifTheSameElseDefault(
-      lines.filter(({ numberOfPacks }) => numberOfPacks > 0),
-      'packSize',
-      -1
-    );
+  const items = lines.map(({ item }) => item);
+  const itemId = ifTheSameElseDefault(items, 'id', '');
+  useEffect(() => setSelected(undefined), [itemId]);
 
-    // on initial selection of an item there will be no packs selected
-    // as the filtering above will have filtered out all the lines as none
-    // will have a number of packs, so instead, look at all the lines except
-    // the placeholder to find either the distinct pack size or use 'any'.
-    if (selectedPackSize === -1)
-      selectedPackSize = ifTheSameElseDefault(
-        lines.filter(
-          ({ type }) => type !== InvoiceLineNodeType.UnallocatedStock
-        ),
-        'packSize',
-        -1
-      );
+  // Effect to auto-select an option. Runs when the item changes (selected is null'd) or
+  // on mounting.
+  useEffect(() => {
+    if (!selected && validPackSizes.length) {
+      // Select the only available pack size.
+      if (validPackSizes.length === 1) {
+        const onlyValidPackSize = validPackSizes[0] as PackSizeOption;
+        setSelected(onlyValidPackSize);
+      }
 
-    setPackSize(selectedPackSize);
-  }, [setPackSize, selected, lines]);
+      // If there are multiple, check if all the allocated lines have the same pack size.
+      // If so, select that pack size. Otherwise, select `any`.
+      if (validPackSizes.length > 1) {
+        const sameAllocated = ifTheSameElseDefault(
+          validPackSizes.filter(({ hasAllocated }) => hasAllocated),
+          'packSize',
+          -1
+        );
+        setPackSize(sameAllocated);
+      }
+    }
+  }, [selected, validPackSizes]);
 
-  return { selected, setPackSize, options, packSizes };
+  return {
+    selected: selected && { label: selected?.label, value: selected?.value },
+    setPackSize,
+    options: validPackSizes.map(({ label, value }) => ({ label, value })),
```

Comment:
```suggestion
    options: validPackSizes
      .sort(sortPackSizes)
      .map(({ label, value }) => ({ label, value })),
```

and this from GitHub copilot 😉  
```
  // Helper to sort the pack sizes by value.
  const sortPackSizes = (a: PackSizeOption, b: PackSizeOption) => {
    if (a.value === b.value) return 0;
    return a.value < b.value ? -1 : 1;
  };
```
a small suggestion, as I've noticed this:

<img width="152" alt="Screen Shot 2022-03-09 at 9 09 38 AM" src="https://user-images.githubusercontent.com/9192912/157320491-a948ced3-3636-4b50-a21c-4db22c4a0e8e.png">

Reply:
Ah yes! Thank you and your copilot!